### PR TITLE
[ONNX][Reland] Export node and value with scope name

### LINKED ISSUE #75100: Preserve original scope names in exported ONNX graph
### 🚀 The feature, motivation and pitch

When a PyTorch model is exported, the graph nodes take uninformative names like `f"{op_type}_{op_cnt}"`, and so do tensors (besides the input and output tensors of the overall graph).
In some contexts, it would be extremely helpful if each node in the exported ONNX graph would be named after the scope name of the PyTorch module from which it came from. Additionally, intermediate tensors should also ideally be named after the node that created them.

Here's a very subjective list of why it matters:

### Visualisation, Debugging

netron_view

Let's say one goal of exporting to ONNX is to visualise a model (without loss of generality, let's think Netron) to reason about its architecture. For anything but the most trivial NN, this is simply not possible if nodes are named `Conv_{N}` and tensors are `{M}`. 

As soon as the node name becomes `unet.encoder.stage1.conv0` and its output tensor `unet.encoder.stage1.conv0.output0`, I become able to understand what node am I looking at, where is a skip connection going, and this sort of things.

### ONNX graph manipulation

For some people, exporting a model to ONNX is not the end of the story. There may be some graph- or weight- level manipulations that must be applied on sub-graphs defined by the name scopes of their nodes or their intermediate tensors: just to give an idea, a silly example here may be "apply batchnorm fusion only on the decoder of the UNet".

### TensorFlow does it 😛 

Let's be clear: I am by no means a TF fan, but credit is due here. On that side, the feature I'm pitching is the default in [`tf2onnx`](https://github.com/onnx/tensorflow-onnx/tree/main/tf2onnx) and is a design decision that makes a lot of sense.

### Notes

I fully understand that name scopes may lose some of their meaning when running graph-level optimisations during the export if sets of nodes are replaced by new ones and there is no 1:1 relationship. Even in this case, though, it may be possible to assign a "common ancestor name"... or fall back to the current implementation if nothing can be done.

### 1. Fork PyTorch and diverge

Since at least some of the code of the ONNX exporting is in C++, there is no easy monkey patching solution, the library must be recompiled.

### 2. Hack around with a heavy hand

[Someone](https://github.com/pytorch/pytorch/issues/33463#issuecomment-606399944) suggested that it is possible to monkey patch the `_slow_forward` method of `torch.nn.Module` to give meaningful names to the _tensors_ flying around during graph tracing (note: not the nodes, but from output tensor names to node names the step is trivial).
I tried it and it seems to be a bit of a hit and miss, I would need to have a way deeper understanding of the guts of JIT to ensure that 100% of the output tensors actually have a useful name.

Alternatively, or in conjunction, I am also trying to monkey patch `torch.onnx.utils._export()` to use the `inlined_graph`, dump it, and rename the ONNX nodes as a post-processing.

### Additional context

The scope name would be readily available from the `scopeName()` of each `torch._C.Node` of the TorchScript's `inlined_graph`, but the export instead uses the `graph` where such information has been stripped off (for example, [here](https://github.com/pytorch/pytorch/blob/00607e75e5f3157839719efb5b26ca92e91190a3/torch/onnx/utils.py#L430) is the first place where the choice of using the `graph` is made and becomes consequential when exporting a `ScriptModule`).

To make use of this metadata down the line, also the serialisation logic of `torch._C.Graph._export_onnx()` must be adapted. [Here](https://github.com/pytorch/pytorch/blob/00607e75e5f3157839719efb5b26ca92e91190a3/torch/csrc/jit/serialization/export.cpp#L830) is where the nodes of the exported ONNX proto take their names on the fly.

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82040

Introduce `_jit_pass_onnx_assign_node_and_value_names` to parse and assign
scoped name for nodes and values in exported onnx graph.
Module layer information is obtained from `ONNXScopeName` captured in `scope`
attribute in nodes. For nodes, the processed onnx node name are stored in
attribute `onnx_name`. For values, the processed onnx output name are stored
as `debugName`.

### CODE REVIEW

File: torch/csrc/onnx/init.cpp [hunk 1]
```diff
@@ -223,6 +223,10 @@ void initONNXBindings(PyObject* module) {
             }
           },
           "Write `args` to the previously specified ONNX log stream.")
+      .def(
+          "_jit_pass_onnx_assign_node_and_value_names",
+          ::torch::jit::onnx::AssignNodeAndValueNames,
```
Comment (justinchuby) [inline torch/csrc/onnx/init.cpp:228]: Use `::torch::wrap_pybind_function` to ensure errors are handled correctly?

**Review by titaiwangms — APPROVED** [review]
:beers:

File: torch/csrc/jit/passes/onnx/naming.cpp [hunk 2]
```diff
@@ -75,6 +75,131 @@ bool isCompatibleScope(torch::jit::ScopePtr scope) {
...
+
+ private:
+  std::string GetFullScopeName(ScopePtr scope);
+  std::unordered_map<ScopePtr, std::string> full_scope_names_;
+  std::unordered_map<std::string, size_t> base_scope_name_counts_;
+};
+
+std::string NodeNameGenerator::UpdateBaseName(
+    std::unordered_map<std::string, size_t>& base_name_count,
+    std::string base_name) {
```
Comment (justinchuby) [inline torch/csrc/jit/passes/onnx/naming.cpp:119]: NOTE: naming

File: torch/csrc/jit/passes/onnx/naming.cpp [hunk 3]
```diff
@@ -75,6 +75,131 @@ bool isCompatibleScope(torch::jit::ScopePtr scope) {
...
+  }
+  n->s_(Symbol::attr(::torch::onnx::OnnxNodeNameAttribute), node_names_[n]);
+}
+
+std::string ScopedNodeNameGenerator::GetFullScopeName(ScopePtr scope) {
+  if (full_scope_names_.find(scope) == full_scope_names_.end()) {
+    auto full_scope_name =
+        ONNXScopeName::variableNameFromRoot(scope, layer_separator_);
+    full_scope_names_[scope] =
+        UpdateBaseName(base_scope_name_counts_, full_scope_name);
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/naming.cpp:191]: TODO: change `UpdateBaseName` function name

File: torch/onnx/utils.py [hunk 4]
```diff
@@ -1279,8 +1257,34 @@ def _track_module_attributes_forward_hook(module, input, output):
             m.register_forward_hook(_track_module_attributes_forward_hook)
             m.register_forward_pre_hook(_track_module_attributes_forward_pre_hook)
 
+    def __unqualified_variable_name(qualified_name: str):
```
Comment (BowenBao) [inline torch/onnx/utils.py:1260]: TODO: move to first PR in stack

File: torch/csrc/onnx/onnx.h [hunk 5]
```diff
@@ -16,5 +16,7 @@ enum class TrainingMode {
   TRAINING, // Training mode
 };
 
+const std::string OnnxNodeNameAttribute = "onnx_name";
```
Comment (BowenBao) [inline torch/csrc/onnx/onnx.h:19]: TODO: k

File: torch/onnx/utils.py (hunk 4 above)
Comment (justinchuby) [inline torch/onnx/utils.py:1260]:
```suggestion
    def __unqualified_variable_name(qualified_name: str) -> str:
```

File: torch/onnx/utils.py [hunk 6]
```diff
@@ -1279,8 +1257,34 @@ def _track_module_attributes_forward_hook(module, input, output):
...
+            >>> __unqualified_variable_name('__main__.Foo.bar')
+            'bar'
+            >>> __unqualified_variable_name('__main__.Foo.bar.0')
+            'bar.0'
+        """
+        name_atoms = qualified_name.split(".")
+        idx = next(
+            (
+                i
+                for i, atom in reversed(list(enumerate(name_atoms)))
```
Comment (BowenBao) [inline torch/onnx/utils.py:1277]:
```python
for i, atom in reversed(list(enumerate(name_atoms))):
    if not atom.isnumeric():
        return ".".join(name_atoms[i:])
return qualified_name
```

**Review by justinchuby — APPROVED** [review]

**Comment by RalphMao** [comment]
This new feature will be super helpful! Looking forward to it!

**Comment by BowenBao** [comment]
@abock for approval

**Review by abock — APPROVED** [review]

**Comment by malfet** [comment]
This change were reverted internally (though it has not propagated to github yet), due to the global constructor in the header, i.e. this line
https://github.com/pytorch/pytorch/blob/df523a6eeef283a79062f828c3fedce2cc3e32f0/torch/csrc/onnx/onnx.h#L19

Global constructors are bad, but global constructors in headers are especially , will take an AI to adjust our linters to make those a fatal error.

P.S. what's wrong with adding it as `const char*`? Or (though I don't know if C++14 standard allows that) `constexpr std::string`?

**Comment by weiwangmeta** [comment]
Manually trigger the revert process trying to resolve sync issues. 
@pytorchbot revert -m="Diff reverted internally" -c="ghfirst"

**Comment by BowenBao** [comment]
@malfet thanks for info, switching to `constexpr char []`.
